File: mobile/src/lib/moderation.ts
/**
 * Reporting a listing and blocking a person.
 *
 * Separate from `listings.ts` because these are not part of the food flow:
 * they are what someone reaches for when the flow has gone wrong, and keeping
 * them apart means the feed code never has to think about moderation.
 */

import { api } from "@/lib/api";

/**
 * Why a listing is being reported.
 *
 * Ordered by how urgently a human should look, not alphabetically — unsafe
 * food is the only one on this list that can put somebody in hospital, so it
 * leads both the type and the picker built from it.
 */
export const REPORT_REASONS = [
	{ value: "unsafe_food", label: "Unsafe or wrongly described food" },
	{ value: "offensive", label: "Offensive or abusive content" },
	{ value: "harassment", label: "Harassment" },
	{ value: "misleading", label: "Misleading or not really available" },
	{ value: "spam", label: "Spam or selling" },
	{ value: "other", label: "Something else" },
] as const;

export type ReportReason = (typeof REPORT_REASONS)[number]["value"];

export interface BlockedUser {
	user_id: string;
	display_name: string | null;
	blocked_at: string;
}

/**
 * Report a listing.
 *
 * Resolves on success and tells the caller nothing about what happened next —
 * the server deliberately does not say whether the listing was removed, since
 * that would make reporting a way to probe other people's accounts.
 *
 * `detail` is required by the server when the reason is `other`; the picker
 * enforces the same rule so the failure never reaches the network.
 */
export async function reportListing(
	listingId: string,
	reason: ReportReason,
	detail?: string,
): Promise<void> {
	await api.post<unknown>(`/listings/${listingId}/report`, {
		reason,
		detail: detail?.trim() ? detail.trim() : undefined,
	});
}

/**
 * Block someone.
 *
 * Their food disappears from your feed and yours from theirs, and neither of
 * you can claim the other's. Blocking twice is a no-op rather than an error,
 * so a double-tap never reads as a failure.
 */
export async function blockUser(userId: string): Promise<void> {
	await api.post<void>(`/users/${userId}/block`);
}

/** Undo a block. A no-op if they were not blocked. */
export async function unblockUser(userId: string): Promise<void> {
	await api.del<void>(`/users/${userId}/block`);
}

/** Who you have blocked, so a block can be undone. Only blocks you made. */
export async function fetchBlockedUsers(): Promise<BlockedUser[]> {
	const body = await api.get<{ items: BlockedUser[]; count: number }>(
		"/users/me/blocks",
	);
	return body.items;
}

File: api/routers/moderation.py
"""Reporting content and blocking people.

Required of any app carrying user-generated content — App Store Guideline 1.2
asks for a way to report offensive material and a way to block abusive users,
and this is it. Kept in one prefix-less router because the two halves hang off
different resources: a report is about a listing, a block is about a person.
"""

import uuid

from fastapi import APIRouter, Depends, status

from api.deps import CurrentUser, DbSession, rate_limited
from api.schemas.moderation import BlockedUser, BlockList, CreateReport, ReportOut
from api.services import moderation as service

router = APIRouter(tags=["moderation"])


@router.post(
    "/listings/{listing_id}/report",
    response_model=ReportOut,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(rate_limited("report", limit=10, window_seconds=60))],
)
async def report_listing(
    listing_id: uuid.UUID,
    body: CreateReport,
    db: DbSession,
    user: CurrentUser,
) -> ReportOut:
    """Report a listing for a human to look at.

    Deliberately succeeds quietly. The reporter is told it was received and
    nothing else — whether the listing was removed, and what happened to the
    account behind it, is not the reporter's to know, and telling them turns
    the feature into a way of probing other people's accounts.
    """
    report = await service.report_listing(
        db,
        listing_id=listing_id,
        reporter_id=user.id,
        reason=body.reason,
        detail=body.detail,
    )
    return ReportOut.model_validate(report)


@router.post("/users/{user_id}/block", status_code=status.HTTP_204_NO_CONTENT)
async def block(user_id: str, db: DbSession, user: CurrentUser) -> None:
    """Block someone.

    Their listings disappear from your feed and yours from theirs, and neither
    of you can claim the other's food. Idempotent: blocking twice is fine.
    """
    await service.block_user(db, blocker_id=user.id, blocked_id=user_id)


@router.delete("/users/{user_id}/block", status_code=status.HTTP_204_NO_CONTENT)
async def unblock(user_id: str, db: DbSession, user: CurrentUser) -> None:
    """Unblock someone. A no-op if they were not blocked."""
    await service.unblock_user(db, blocker_id=user.id, blocked_id=user_id)


@router.get("/users/me/blocks", response_model=BlockList)
async def my_blocks(db: DbSession, user: CurrentUser) -> BlockList:
    """Who you have blocked, so a block can be undone.

    Only blocks you made. Being blocked by somebody is not shown — a block is
    meant to be a quiet exit, and surfacing it would make it a confrontation.
    """
    rows = await service.list_blocks(db, user.id)
    return BlockList(
        items=[
            BlockedUser(
                user_id=other.id,
                display_name=other.name,
                blocked_at=block.created_at,
            )
            for other, block in rows
        ],
        count=len(rows),
    )
